pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,572 @@
+                                        recvBuf.data(), sendCount, datatype,
+                                        this->getComm()));
+        return recvBuf;
+    }
+
+    /**
+     * @brief MPI_Allgather wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer vector.
+     * @param recvBuf Receive buffer vector.
+     */
+    template <typename T>
+    void Allgather(std::vector<T> &sendBuf, std::vector<T> &recvBuf) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        PL_ABORT_IF(recvBuf.size() != sendBuf.size() * this->getSize(),
+                    "Incompatible size of sendBuf and recvBuf.");
+        PL_MPI_IS_SUCCESS(MPI_Allgather(
+            sendBuf.data(), sendBuf.size(), datatype, recvBuf.data(),
+            sendBuf.size(), datatype, this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Allgather wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer vector.
+     *
+     * @return recvBuf Vector of receive buffer.
+     */
+    template <typename T>
+    auto allgather(std::vector<T> &sendBuf) -> std::vector<T> {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        std::vector<T> recvBuf(sendBuf.size() * this->getSize());
+        PL_MPI_IS_SUCCESS(MPI_Allgather(
+            sendBuf.data(), sendBuf.size(), datatype, recvBuf.data(),
+            sendBuf.size(), datatype, this->getComm()));
+        return recvBuf;
+    }
+
+    /**
+     * @brief MPI_Allreduce wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer.
+     * @param recvBuf Receive buffer.
+     * @param op_str String of MPI_Op.
+     */
+    template <typename T>
+    void Allreduce(T &sendBuf, T &recvBuf, const std::string &op_str) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Op op = getMPIOpType(op_str);
+        PL_MPI_IS_SUCCESS(MPI_Allreduce(&sendBuf, &recvBuf, 1, datatype, op,
+                                        this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Allreduce wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer.
+     * @param op_str String of MPI_Op.
+     *
+     * @return recvBuf Receive buffer.
+     */
+    template <typename T>
+    auto allreduce(T &sendBuf, const std::string &op_str) -> T {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Op op = getMPIOpType(op_str);
+        T recvBuf;
+        PL_MPI_IS_SUCCESS(MPI_Allreduce(&sendBuf, &recvBuf, 1, datatype, op,
+                                        this->getComm()));
+        return recvBuf;
+    }
+
+    /**
+     * @brief MPI_Allreduce wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer vector.
+     * @param recvBuf Receive buffer vector.
+     * @param op_str String of MPI_Op.
+     */
+    template <typename T>
+    void Allreduce(std::vector<T> &sendBuf, std::vector<T> &recvBuf,
```

Comment:
Should we validate both vectors have the same size?

Reply:
Good point!